demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
+        Each state supports common methods like get_status_message(), get_status_color(), allow_configuration_changes()
+        """
+        return self._state
+
+    def build_form():
+        """ Convert self.configuration_schema into a Django form """
+        pass
+
+
+class VmProvider(Resource):
+    """ A provider that can provision virtual machines """
+    class Meta:
+        proxy = True
+
+    def get_flavors(self):
+        """ Get available VM sizes """
+        raise NotImplementedError
+    def list_instances(self):
+        """ List instances currently provisioned on this account. """
+        raise NotImplementedError
+    def provision_vm(self, flavor, image):
+        """ Start the process of provisioning a new VM. Returns a VmInstance """
+        raise NotImplementedError
+
+
+class OpenStackVmProvider(VmProvider):
+    # Implementation here.
+
+
+class VmInstance(Resource):
+    """
+    A virtual machine.
+    configuration options: flavor, image, is_powered_on
+    """
+    class Meta:
+        proxy = True
+
+
+class OpenStackVmInstance(VmInstance):
+    # Implementation here
+
+
+class OpenEdxSandboxInstance(Resource):
```

Comment:
It would probably be an independent class, with a mixin used to implement common functionality. There would be quite a number of differences I suspect.